mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala

```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+   * negative samples.
+   */
+  private def generateUnigramTable(normalizedWeights: Array[Double], tableSize: Int): Array[Int] = {
+    val table = Array.fill(tableSize)(0)
+    var a = 0
+    var i = 0
+    while (a < table.length) {
+      table.update(a, i)
+      if (a.toFloat / table.length >= normalizedWeights(i)) {
+        i = math.min(normalizedWeights.length - 1, i + 1)
+      }
+      a += 1
+    }
+    table
+  }
+
+  private def generateVocab[S <: Iterable[String]](input: RDD[S]):
+      (Int, Long, Map[String, Int], Array[Int]) = {
+    val sc = input.context
+
+    val words = input.flatMap(x => x)
+
+    val vocab = words.map(w => (w, 1L))
+      .reduceByKey(_ + _)
+      .filter{case (w, c) => c >= $(minCount)}
+      .collect()
+      .sortWith{case ((w1, c1), (w2, c2)) => c1 > c2}
+
+    val totalWordCount = vocab.map(_._2).sum
+
+    val vocabMap = vocab.zipWithIndex.map{case ((w, c), i) =>
+      w -> i
+    }.toMap
+
+    // We create a cumulative distribution array, unlike the original implemention
+    // and use binary search to get insertion points. This should replicate the same
+    // behavior as the table in original implementation.
+    val weights = vocab.map(x => scala.math.pow(x._2, power))
+    val totalWeight = weights.sum
+
+    val normalizedCumWeights = weights.scanLeft(0.0)(_ + _).tail.map(x => (x / totalWeight))
+
+    val unigramTableSize =
+      math.min(maxUnigramTableSize, unigramTableSizeFactor * normalizedCumWeights.length)
+    val unigramTable = generateUnigramTable(normalizedCumWeights, unigramTableSize)
+
+    (vocabMap.size, totalWordCount, vocabMap, unigramTable)
+  }
+
+  /**
+   * This method implements Word2Vec Continuous Bag Of Words based implementation using
+   * negative sampling optimization, using BLAS for vectorizing operations where applicable.
+   * The algorithm is parallelized in the same way as the skip-gram based estimation.
+   * @param input
+   * @return
+   */
+  private def fitCBOW[S <: Iterable[String]](input: RDD[S]): feature.Word2VecModel = {
+    val (vocabSize, totalWordCount, vocabMap, uniTable) = generateVocab(input)
+    val negSamples = $(negativeSamples)
+    assert(negSamples < vocabSize,
+      s"Vocab size ($vocabSize) cannot be smaller than negative samples($negSamples)")
+    val seed = $(this.seed)
+    val initRandom = new XORShiftRandom(seed)
+
+    val vectorSize = $(this.vectorSize)
+
+    val syn0Global = Array.fill(vocabSize * vectorSize)(initRandom.nextFloat - 0.5f)
+    val syn1Global = Array.fill(vocabSize * vectorSize)(0.0f)
+
+    val sc = input.context
+
+    val vocabMapbc = sc.broadcast(vocabMap)
+    val unigramTablebc = sc.broadcast(uniTable)
+
+    val window = $(windowSize)
+
+    val digitSentences = input.flatMap{sentence =>
+      val wordIndexes = sentence.flatMap(vocabMapbc.value.get)
+      wordIndexes.grouped($(maxSentenceLength)).map(_.toArray)
```

Comment:
If we wrap up the algo in a separate object, and feed in the `Param` values as args, we should only capture the object scope and not need to bc all the little params?